### PR TITLE
Add support for Secp256k1 signatures

### PR DESCRIPTION
**Pull request by romac**
Close: #939 

- [x] Add support for verifying Secp256k1 signatures.
- [x] Turn the `Signature` enum into a wrapper over a byte array, so that it can hold signatures of any type. As the `Signature` is only ever used/useful in conjunction with a `PublicKey`, the latter is enough to determine which type of signature it holds.
- [x] Use `Option<Signature>` instead of `Signature::None` in various places.
- [x] Write tests using the test vectors provided below in the comments
- [x] Add a length check to the `Signature` constructor

---

* [x] Referenced an issue explaining the need for the change
* [ ] Updated all relevant documentation in docs
* [ ] Updated all code comments where relevant
* [x] Wrote tests
* [x] Added entry in `.changelog/`

### CODE REVIEW

File: tendermint/src/signature.rs [hunk 1]
```diff
@@ -4,103 +4,73 @@ pub use ed25519::{Signature as Ed25519Signature, SIGNATURE_LENGTH as ED25519_SIG
...
-    /// No signature present
-    None, /* This could have been implemented as an `Option<>` but then handling it would be
-           * outside the scope of this enum. */
-}
+#[derive(Clone, Debug, PartialEq)]
+pub struct Signature(Vec<u8>);
 
 impl Protobuf<Vec<u8>> for Signature {}
 
 impl TryFrom<Vec<u8>> for Signature {
```
[the remark this answers is not in the document]
Comment (romac) [reply tendermint/src/signature.rs:20]: The `TryFrom<Vec<u8>>` instance is required by the `Protobuf<Vec<u8>>` instance, but I agree that we should add a `TryFrom<&[u8]>` instance as well.

```rust
pub trait Protobuf<T: Message + From<Self> + Default>
where
    Self: Sized + Clone + TryFrom<T>, // <---------
    <Self as TryFrom<T>>::Error: Display,
```

File: tendermint/src/signature.rs [hunk 2]
```diff
@@ -4,103 +4,73 @@ pub use ed25519::{Signature as Ed25519Signature, SIGNATURE_LENGTH as ED25519_SIG
...
-#[non_exhaustive]
-pub enum Signature {
-    /// Ed25519 block signature
-    Ed25519(Ed25519Signature),
-    /// No signature present
-    None, /* This could have been implemented as an `Option<>` but then handling it would be
-           * outside the scope of this enum. */
-}
+#[derive(Clone, Debug, PartialEq)]
+pub struct Signature(Vec<u8>);
```
[the remark this answers is not in the document]
Comment (tony-iqlusion) [reply tendermint/src/signature.rs:16]: `TryFrom<&[u8]>` will also accomplish the same thing, and is a bit simpler.

File: tendermint/src/public_key.rs [hunk 3]
```diff
@@ -130,16 +130,33 @@ impl PublicKey {
...
+                        Error::signature_invalid(
+                            "Secp256k1 signature verification failed".to_string(),
+                        )
+                    }),
+                    Err(e) => Err(Error::signature_invalid(format!(
+                        "invalid Secp256k1 signature: {}",
+                        e
+                    ))),
+                }
+            }
```
[the remark this answers is not in the document]
Comment (thanethomson) [reply tendermint/src/public_key.rs:159]: I've tried adding a few test vectors from RFC 8032 and the Wycheproof test suite in #964.

**Comment by thanethomson** [comment]
@tony-iqlusion since @romac is on vacation at the moment, I'll make updates to this PR in #964. Once we're happy with #964, we can merge into this PR and subsequently into `master`.

**Review by tony-iqlusion — APPROVED** [review]

**Pull request merged**